### PR TITLE
option to exclude some nodes (i.e. the faulty nodes) from the computa…

### PR DESCRIPTION
**Pull request by hassanzadeh**
Right now, the wrapper assigns the jobs to all available nodes. This causes the pipeline to break if some nodes are faulty. By adding an additional option "exclude_list" to the param file you can exclude some nodes (exclude_list: eg_node_0,eg_node_1,eg_node_3).

### CODE REVIEW

**Comment by marjorief** [comment]
@LichMaster98 : do you have time to look at this?

**Comment by LichMaster98** [comment]
@marjorief I do yes.

**Comment by joecummings** [comment]
I think this needs to be added to `unify()`

File: pegasus_wrapper/resource_request.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ class SlurmResourceRequest(ResourceRequest):
         default=_DEFAULT_JOB_TIME_IN_MINUTES,
         kw_only=True,
     )
+    exclude_list: Optional[str] = attrib(validator=optional(instance_of(str)), kw_only=True, default=None)
```
Comment (LichMaster98) [inline pegasus_wrapper/resource_request.py:92]: Can you add a comment explaining how the `exclude_list` should be formatted in the params file?

File: pegasus_wrapper/resource_request.py [hunk 2]
```diff
@@ -100,6 +101,7 @@ def from_parameters(params: Parameters) -> ResourceRequest:
             if "memory" in params
             else None,
             job_time_in_minutes=params.optional_integer("job_time_in_minutes"),
+            exclude_list=params.optional_string("exclude_list")
         )
 
     def unify(self, other: ResourceRequest) -> ResourceRequest:
```
Comment (LichMaster98) [inline pegasus_wrapper/resource_request.py:107]: Mirroring a comment from @joecummings, this new variable is missing from unify

File: pegasus_wrapper/resource_request.py [hunk 3]
```diff
@@ -146,6 +148,10 @@ def apply_to_job(self, job: Job, *, job_name: str) -> None:
                 else _DEFAULT_JOB_TIME_IN_MINUTES
             ),
         )
+
+        if exclude_list:
```
Comment (LichMaster98) [inline pegasus_wrapper/resource_request.py:152]: Can you add a unit test which tests having nodes excluded in a test workflow?

Comment (hassanzadeh) [reply pegasus_wrapper/resource_request.py:152]: Yes.

**Comment by joecummings** [comment]
Also, might make sense to be able to specify nodes you DO want to run on as a follow-up PR. I know sometimes I want to specify a node that I'm sure has the right specifications.

File: README.md [hunk 4]
```diff
@@ -60,17 +68,9 @@ The experiment directory can be (and ought to be) on such a drive, though.
...
 # Contributing
 
 Run `make precommit` before commiting.  
 
 If you are using PyCharm, please set your docstring format to "Google" and your unit test runner to "PyTest"
 in `Preferences | Tools | Python Integrated Tools`.
-
-## Debugging Tricks/Tips
-
-* Investigate reasons for why a job is in a certain state (like being held for way too long): `condor_q -long [job_id]`. You can get the job id by `pegasus-status -v [wfdir]`.
```
Comment (LichMaster98) [inline README.md:76]: Why are these two sections of the readme being removed?

File: pegasus_wrapper/resource_request.py [hunk 5]
```diff
@@ -116,6 +129,8 @@ def unify(self, other: ResourceRequest) -> ResourceRequest:
             job_time_in_minutes=other.job_time_in_minutes
             if other.job_time_in_minutes
             else self.job_time_in_minutes,
+            exclude_list=other.exclude_list,
+            run_on_single_node=other.run_on_single_node,
```
Comment (LichMaster98) [inline pegasus_wrapper/resource_request.py:133]: It looks like both of these are missing a conditional check to pick from either `self` or `other`. 
```suggestion
            exclude_list=other.exclude_list if other.exclude_list else self.exclude_list,
            run_on_single_node=other.run_on_single_node if other.run_on_single_node else self.run_on_single_node,
```

File: pegasus_wrapper/resource_request.py [hunk 6]
```diff
@@ -125,6 +140,8 @@ def convert_time_to_slurm_format(self, job_time_in_minutes: int) -> str:
     def apply_to_job(self, job: Job, *, job_name: str) -> None:
         if not self.partition:
             raise RuntimeError("A partition to run on must be specified.")
+        exclude_list = self.exclude_list
+        run_on_single_node = self.run_on_single_node
```
Comment (LichMaster98) [inline pegasus_wrapper/resource_request.py:144]: nit: I'm unsure why these need to be assigned to local variables before being used further down in the code?

File: README.md (hunk 4 above)
Comment (hassanzadeh) [reply README.md:76]: Oh, that was removed unintentionally. I'm returning it.

File: pegasus_wrapper/resource_request.py (hunk 5 above)
Comment (LichMaster98) [reply pegasus_wrapper/resource_request.py:133]: @hassanzadeh This comment still appears to be outstanding as an issue to resolve.

File: pegasus_wrapper/resource_request.py (hunk 6 above)
Comment (LichMaster98) [reply pegasus_wrapper/resource_request.py:144]: @hassanzadeh Is there a specific reason you were assigning these to local variables?

**Pull request merged**